crates/core/app/tests/app_can_undelegate_from_a_validator.rs

```diff
@@ -0,0 +1,374 @@
+            /*0, TODO(kate): we still see delegation tokens after undelegating*/ 1,
+            "client should not have delegation tokens immediately after undelegating"
+        );
+    }
+
+    // Jump to the end of the unbonding period.
+    {
+        let jump_to = expected_end_of_unboding_period_height;
+        while get_latest_height().await < jump_to {
+            node.block().execute().await?;
+        }
+    }
+
+    // Build a transaction that will now reclaim staking tokens from the validator.
+    let plan = {
+        client.sync_to_latest(storage.latest_snapshot()).await?;
+        let penalty = penumbra_stake::Penalty::from_percent(0);
+        let note = client
+            .notes
+            .values()
+            .cloned()
+            .filter(|n| n.asset_id() == undelegate_token_id)
+            .next()
+            .expect("should have an unbonding note");
+        let claim = UndelegateClaimPlan {
+            validator_identity: identity_key,
+            unbonding_start_height: pre_undelegated_epoch.start_height,
+            penalty,
+            unbonding_amount: note.amount(),
+            balance_blinding: decaf377::Fr::rand(&mut OsRng),
+            proof_blinding_r: decaf377::Fq::rand(&mut OsRng),
+            proof_blinding_s: decaf377::Fq::rand(&mut OsRng),
+        };
+        let mut plan = TransactionPlan {
+            actions: vec![claim.into()],
+            // Now fill out the remaining parts of the transaction needed for verification:
+            memo: MemoPlan::new(&mut OsRng, MemoPlaintext::blank_memo(*test_keys::ADDRESS_0))
+                .map(Some)?,
+            detection_data: None, // We'll set this automatically below
+            transaction_parameters: TransactionParameters {
+                chain_id: TestNode::<()>::CHAIN_ID.to_string(),
+                ..Default::default()
+            },
+        };
+        plan.populate_detection_data(rand_core::OsRng, 0);
+        plan
+    };
+    let tx = client.witness_auth_build(&plan).await?;
+
+    // Execute the transaction, applying it to the chain state.
+    node.block()
+        .add_tx(tx.encode_to_vec())
+        .execute()
+        .instrument(error_span!("executing block with undelegation claim"))
+        .await?;
+    let post_claim_snapshot = storage.latest_snapshot();
+
+    {
+        client.sync_to_latest(post_claim_snapshot.clone()).await?;
+        assert_eq!(
+            client
+                .notes_by_asset(*penumbra_asset::STAKING_TOKEN_ASSET_ID)
+                .count(),
+            1,
+            "client should still have staking notes"
+        );
+        assert_eq!(
+            client.notes_by_asset(undelegate_token_id).count(),
+            1,
+            "client should still have undelegation notes"
+        );
+        assert_eq!(
+            client.notes_by_asset(delegate_token_id).count(),
+            1,
+            "client should still have delegation notes"
+        );
```

Comment:
Just two ideas, it would be good to check:
- the rate at which delegation <> undelegation token gets converted (by dividing the amounts and comparing it against the stated validator rate)
- the rate at which undelegation <> staking token gets converted (so that later we can pull that logic to check that slashing events effect that rate correctly)

Reply:
i've pushed 84516c6c34bec87ba85542ba007c1690e283ac58 as an initial step in that direction. i have read through the protocol specification, but i am still not very familiar with the rates/arithmetic involved here.

would you be available to pair on this sometime tomorrow (the 28th) to step through this?

Reply:
Yeah for sure!

Reply:
okay we now have some additional assertions to check that validator rates look correct :tada: